Add documentation for AsyncEmbeddedEngine (#5435)
debezium/debezium: documentation/modules/ROOT/pages/development/engine.adoc

This `debezium-api` module defines a small API that allows an application to easily configure and run {prodname} connectors using {prodname} Engine.

Implementation-wise, starting 2.6.0 release, {prodname} provides two implementations of `DebeziumEngine` interface.

> The older one, `EmbeddedEngine`, processes all records sequentially and allows to run on one task for given connector.

Reviewer: ok, got it. Updated my suggested edit.

This the default implementation.
Beginning with the 2.6.0 release, a new `AsyncEmbeddedEngine` implementation is available.

> This implementation processes records in multiple threads, and can run multiple tasks, if the connector supports it (currently only the connectors for SQL Server and MongoDB support running multiple tasks within a single connector).

Reviewer: ```suggestion
This implementation also runs only a single connector, but it can process records in multiple threads, and run multiple tasks, if the connector supports it (currently only the connectors for SQL Server and MongoDB support running multiple tasks within a single connector).
```

This `debezium-api` module defines a small API that allows an application to easily configure and run {prodname} connectors using {prodname} Engine.

Beginning with the 2.6.0 release, {prodname} provides two implementations of the `DebeziumEngine` interface.
The older `EmbeddedEngine` implementation runs a single connector that uses only one task.
The connector emits all records sequentially.

>  This the default implementation.

Reviewer: ```suggestion
This the default implementation.
```
Extra leading space.

> The connector emits all records sequentially.
>  
> This the default implementation.
> Beginning with the 2.6.0 release, a new `AsyncEmbeddedEngine` implementation is available.

Reviewer: ```suggestion
The connector emits all records sequentially.
 This the default implementation.
 
Beginning with the 2.6.0 release, a new `AsyncEmbeddedEngine` implementation is available.
```
Probably should remove the blank line on L26, and then start the new paragraph after that by adding a blank line after the sentence that begins _This is the default ..._

This implementation also runs only a single connector, but it can process records in multiple threads, and run multiple tasks, if the connector supports it (currently only the connectors for SQL Server and MongoDB support running multiple tasks within a single connector).
Because both of these engines implement the same interface and share the same API, the code examples that follow are valid for either engine.

> Both implementations also support the same configuration options.

Reviewer: ```suggestion
Both implementations support the same configuration options.
```

> However, the new `AsyncEmbeddedEngine` provides a couple of new configuration options for setting up and fine-tuning parallel processing.

Reviewer: ```suggestion

However, the new `AsyncEmbeddedEngine` provides a couple of new configuration options for setting up and fine-tuning parallel processing.
```
Since I'm in here to make the previous suggestion, I'll add that you should start a new paragraph with this sentence. :)

Your application should always properly stop the engine to ensure graceful and complete shutdown and that each source record is sent to the application exactly one time.
For example, do not rely upon shutting down the `ExecutorService`, since that interrupts the running threads.
Although the `DebeziumEngine` will indeed terminate when its thread is interrupted, the engine may not terminate cleanly, and when your application is restarted it may see some of the same source records that it had processed just prior to the shutdown.

As mention above, there are two implementations of `DebeziumEngine` interface.
Both use the same API and the code sample above are valid for both of them.
The only exception is creation of the `DebeziumEngine` instance.
As also mentioned in the intorduction, by default `EmbeddedEgine` is used.
Therefore `DebeziumEngine.create(Json.class)` would result in using `EmbeddedEngine` instance under the hood.
If you want to use new `AsyncEmbeddedEngine` instance instead, you have to use `DebeziumEngine#create(KeyValueHeaderChangeEventFormat<K, V, H> format, String builderFactory)` method.

> E.g. to create embedded engine which uses `AsyncEmbeddedEngine` under the hood and uses JSON as a key, value and header format, you need to use following code:

Reviewer: ```suggestion
For example, to create an embedded engine that uses the `AsyncEmbeddedEngine`, and uses JSON as its key, value, and header format, you would use following code:
```

Alternatively you can register `CompletionCallback` while creating `DebeziumEngine` as a callback to be informed when the engine terminates.

Recall that when the JVM shuts down, it only waits for non-daemon threads.
Therefore, when you run the engine on a daemon thread, if your application exits, be sure to wait for the engine process to complete.

Your application should always properly stop the engine to ensure graceful and complete shutdown and that each source record is sent to the application exactly one time.
For example, do not rely upon shutting down the `ExecutorService`, since that interrupts the running threads.
Although the `DebeziumEngine` will indeed terminate when its thread is interrupted, the engine may not terminate cleanly, and when your application is restarted it may see some of the same source records that it had processed just prior to the shutdown.

As mentioned earlier, there are two implementations of the `DebeziumEngine` interface.

> The two implementations use the same API and the preceding code sample is valid for both versions.

Reviewer: ```suggestion
The two implementations use the same API, and the preceding code sample is valid for both versions.
```

* `Json.class` - the header values are encoded as `JSON` strings
* `JsonByteArray.class` - the header values are formatted as `JSON` and encoded as UTF-8 byte arrays

> Internally, the engine uses the appropriate Kafka Connect or Apicurio converter implementation to which the conversion is delegated.

Reviewer: ```suggestion
Internally, the engine delegates data conversion to the Kafka Connect or Apicurio converter implementation with the algorithm that is best suited to perform the conversion.
```

|Max delay (in ms) between retries when encountering connection errors.
|===

[[async-engine-properties]]
=== Asynchronous Engine Properties

[cols="35%a,10%a,55%a"]
|===
|Property
|Default
|Description

|`record.processing.threads`

> |`#` of availabel cores

Reviewer: ```suggestion
|The number of available cores.
```

> |The number of threads to be used for processing CDC records. The default is number of available machine cores with upper limit of 16 threads.

Reviewer: ```suggestion
|The number of threads to use for processing CDC records. 
The default is the number of available machine cores, with an upper limit of 16 threads.
```

> If you want to use all available thread without any limitation, use `AVAILABLE_CORES` placeholder.

Reviewer: ```suggestion
If you want to use all available threads without any limitation, use the `AVAILABLE_CORES` placeholder.
```

|`record.processing.shutdown.timeout.ms`
|1000

> |Maximum time in milliseconds to wait for processing submitted records when task shutdown is called.

Reviewer: ```suggestion
|Maximum time in milliseconds to wait for processing submitted records after a task shutdown is called.
```

|`record.processing.order`
|`ORDERED`
|Determines how the records should be produced.

> * `ORDERED` means sequential processing, i.e. that the records are produced in the same order as they were obtained from the database.
> * `UNORDERED` means non-sequential processing, i.e. the records can be produced in a different order than the original one.

Reviewer: ```suggestion
`ORDERED`:: Records are processed sequentially; that is, they are produced in the order in which they were obtained from the database.
`UNORDERED`:: Records are processed non-sequentially; that is, they can be produced in an different order than in the source database.
```

> Non-sequential approach gives better throughput, as the records are produced immediately once the SMTs and serialization of the message is done, without waiting of other records.

Reviewer: ```suggestion
The non-sequential processing of the `UNORDERED` option results in better throughput, because records are produced immediately after any SMT processing and message serialization is complete, without waiting for other records.
```

> This option doesn't have any effect when ChangeConsumer is provided to the engine.

Reviewer: ```suggestion
This option doesn't have any effect when the `ChangeConsumer` method is provided to the engine.
```

|`record.processing.with.serial.consumer`
|`false`

> |Specifies whether the default ChangeConsumer should be created from provided Consumer, resulting in serial Consumer processing.

Reviewer: ```suggestion
|Specifies whether the default `ChangeConsumer` should be created from the provided `Consumer`, resulting in serial `Consumer` processing.
```

> This option has no effect if the ChangeConsumer is already provided to the engine via configuration.

Reviewer: ```suggestion
This option has no effect if the `ChangeConsumer` is already provided to the engine via configuration.
```

|`task.management.timeout.ms`
|180,000 (3 min)

> |Time to wait for task's lifecycle management operations (starting and stopping), given in milliseconds.

Reviewer: ```suggestion
|Time, in milliseconds, that the engine waits for a task's lifecycle management operations (starting and stopping) to complete.
```
What happens if the specified time is exceeded?

Author: Engine throws an exception, which results in stopping the engine, so if the time is exceed during start, engine fails to start, in case of shutdown, engine would be stopped abruptly (e.g. before processing all buffered records, not storing latest offset etc.)
